### PR TITLE
Port StringType

### PR DESCRIPTION
**Pull request by cston**
Port `StringType` from reference source.

### CODE REVIEW

**Comment by cston** [comment]
cc @danmosemsft, @jaredpar, @KathleenDollard

File: src/Microsoft.VisualBasic.Core/src/Microsoft/VisualBasic/CompilerServices/CharArrayType.vb [hunk 1]
```diff
@@ -0,0 +1,65 @@
...
+    Public NotInheritable Class CharArrayType
+        ' Prevent creation.
+        Private Sub New()
+        End Sub
+
+        Public Shared Function FromString(ByVal Value As String) As Char()
+
+            If Value Is Nothing Then
+
+                Value = ""
```
Comment (am11) [inline src/Microsoft.VisualBasic.Core/src/Microsoft/VisualBasic/CompilerServices/CharArrayType.vb]: Maybe just `Return Array.Empty<Char>()`, here and on line 32?

**Comment by cston** [comment]
@333fred, @anipik, please review, thanks.

File: src/Microsoft.VisualBasic.Core/tests/ConversionsTests.cs [hunk 2]
```diff
@@ -2552,6 +2552,179 @@ public void ToChar_InvalidBool_ThrowsInvalidOperationException(object value)
...
+        public static IEnumerable<object[]> ToString_Object_TestData()
+        {
+            // char[]
+            yield return new object[] { new char[0], "" };
+            yield return new object[] { new char[] { (char)0 }, "\0" };
+            yield return new object[] { new char[] { 'A', 'B' }, "AB" };
+        }
+
+        [Theory]
+        [MemberData(nameof(ToString_Object_TestData))]
```
Comment (Anipik) [inline src/Microsoft.VisualBasic.Core/tests/ConversionsTests.cs]: I think you can add ```[MemberData(nameof(ToString_IConvertible_TestData))]``` as well here

File: src/Microsoft.VisualBasic.Core/tests/ConversionsTests.cs [hunk 3]
```diff
@@ -2552,6 +2552,179 @@ public void ToChar_InvalidBool_ThrowsInvalidOperationException(object value)
...
+                AssertEqual(expected, Conversions.ToString(new ConvertibleWrapper(value)));
+            }
+        }
+
+        public static IEnumerable<object[]> ToString_Object_TestData()
+        {
+            // char[]
+            yield return new object[] { new char[0], "" };
+            yield return new object[] { new char[] { (char)0 }, "\0" };
+            yield return new object[] { new char[] { 'A', 'B' }, "AB" };
```
Comment (Anipik) [inline src/Microsoft.VisualBasic.Core/tests/ConversionsTests.cs]: why  this data cant be added to ```ToString_IConvertible_TestData``` method ?

File: src/Microsoft.VisualBasic.Core/src/Microsoft.VisualBasic.Core.vbproj [hunk 4]
```diff
@@ -28,16 +28,19 @@
     <Compile Include="Microsoft\VisualBasic\FileIO\FileSystem.Unix.vb" />
   </ItemGroup>
   <ItemGroup>
-    <Compile Include="Microsoft\VisualBasic\CallType.vb" />
-    <Compile Include="Microsoft\VisualBasic\Collection.vb" />
     <Compile Include="Microsoft\VisualBasic\ApplicationServices\StartupEventArgs.vb" />
     <Compile Include="Microsoft\VisualBasic\ApplicationServices\StartupNextInstanceEventArgs.vb" />
     <Compile Include="Microsoft\VisualBasic\ApplicationServices\UnhandledExceptionEventArgs.vb" />
+    <Compile Include="Microsoft\VisualBasic\CallType.vb" />
```
Comment (333fred) [inline src/Microsoft.VisualBasic.Core/src/Microsoft.VisualBasic.Core.vbproj]: Are we able to just use globs here?

File: src/Microsoft.VisualBasic.Core/tests/ConversionsTests.cs [hunk 5]
```diff
@@ -2552,6 +2552,179 @@ public void ToChar_InvalidBool_ThrowsInvalidOperationException(object value)
             Assert.Throws<InvalidOperationException>(() => Conversions.ToChar(value));
         }
 
+        public static IEnumerable<object[]> ToString_IConvertible_TestData()
```
Comment (Anipik) [inline src/Microsoft.VisualBasic.Core/tests/ConversionsTests.cs]: This data is declared twice, we can just use data from stringTypeTests.cs

File: src/Microsoft.VisualBasic.Core/tests/StringTypeTests.cs [hunk 6]
```diff
@@ -0,0 +1,386 @@
...
+        [InlineData("abc", 2, 3, "def", "ade")]
+        public void MidStmtStr(string str, int start, int length, string insert, string expected)
+        {
+            StringType.MidStmtStr(ref str, start, length, insert);
+            Assert.Equal(expected, str);
+        }
+
+        [Theory]
+        [InlineData("", -1, 0, "")]
+        [InlineData("", 0, 0, "")]
```
Comment (Anipik) [inline src/Microsoft.VisualBasic.Core/tests/StringTypeTests.cs]: can you add above 2 testcases for non empty strings as well

File: src/Microsoft.VisualBasic.Core/tests/StringTypeTests.cs [hunk 7]
```diff
@@ -0,0 +1,386 @@
...
+        [InlineData("", 1, 0, "")]
+        [InlineData("", 2, 0, "")]
+        [InlineData("", 1, -1, "")]
+        [InlineData("abc", 4, 0, "")]
+        [InlineData("abc", 1, -3, "")]
```
Comment (Anipik) [inline src/Microsoft.VisualBasic.Core/tests/StringTypeTests.cs]: can you add test cases for null input as well

File: src/Microsoft.VisualBasic.Core/tests/StringTypeTests.cs [hunk 8]
```diff
@@ -0,0 +1,386 @@
...
+        }
+
+        [Theory]
+        [InlineData("a", "a", 0, 0)]
+        [InlineData("a", "b", -1, -1)]
+        [InlineData("b", "a", 1, 1)]
+        [InlineData("a", "ABC", 32, -1)]
+        [InlineData("ABC", "a", -32, 1)]
+        [InlineData("abc", "ABC", 32, 0)]
+        public void StrCmp(string left, string right, int expectedBinaryCompare, int expectedTextCompare)
```
Comment (Anipik) [inline src/Microsoft.VisualBasic.Core/tests/StringTypeTests.cs]: can we have some more test cases with different lengths

File: src/Microsoft.VisualBasic.Core/tests/StringTypeTests.cs [hunk 9]
```diff
@@ -0,0 +1,386 @@
...
+        [InlineData("abc", 1, -3, "")]
+        public void MidStmtStr_ArgumentException(string str, int start, int length, string insert)
+        {
+            Assert.Throws<ArgumentException>(() => StringType.MidStmtStr(ref str, start, length, insert));
+        }
+
+        [Theory]
+        [InlineData("a", "a", 0, 0)]
+        [InlineData("a", "b", -1, -1)]
+        [InlineData("b", "a", 1, 1)]
```
Comment (Anipik) [inline src/Microsoft.VisualBasic.Core/tests/StringTypeTests.cs]: tests cases if the input is null or empty

File: src/Microsoft.VisualBasic.Core/tests/StringTypeTests.cs [hunk 10]
```diff
@@ -0,0 +1,386 @@
...
+            Assert.Equal(expectedBinaryCompare, StringType.StrCmp(left, right, TextCompare: false));
+            Assert.Equal(expectedTextCompare, StringType.StrCmp(left, right, TextCompare: true));
+        }
+
+        [Theory]
+        [InlineData("", "*", true, true)]
+        [InlineData("", "?", false, false)]
+        [InlineData("a", "?", true, true)]
+        [InlineData("a3", "[A-Z]#", false, true)]
+        [InlineData("A3", "[a-z]#", false, true)]
```
Comment (Anipik) [inline src/Microsoft.VisualBasic.Core/tests/StringTypeTests.cs]: can you add some test cases with more patterns and strings. Also tests for null inputs are missings here

File: src/Microsoft.VisualBasic.Core/src/Microsoft.VisualBasic.Core.vbproj (hunk 4 above)
Comment (danmosemsft) [reply src/Microsoft.VisualBasic.Core/src/Microsoft.VisualBasic.Core.vbproj]: In CoreFX we don't use them, basically only for test assets. But my care level is low (as long as it's "*.vb" and not "*")

File: src/Microsoft.VisualBasic.Core/tests/ConversionsTests.cs [hunk 11]
```diff
@@ -2552,6 +2552,179 @@ public void ToChar_InvalidBool_ThrowsInvalidOperationException(object value)
...
+            yield return new object[] { uint.MinValue, "0" };
+            yield return new object[] { (uint)1, "1" };
+            yield return new object[] { uint.MaxValue, "4294967295" };
+            yield return new object[] { (UIntEnum)uint.MinValue, "0" };
+            yield return new object[] { (UIntEnum)1, "1" };
+            yield return new object[] { (UIntEnum)uint.MaxValue, "4294967295" };
+
+            // int.
+            yield return new object[] { int.MinValue, "-2147483648" };
+            yield return new object[] { -1, "-1" };
```
[the remark this answers is not in the document]
Comment (cston) [reply src/Microsoft.VisualBasic.Core/tests/ConversionsTests.cs]: Already covered by `int.MinValue` and `int.MaxValue`.

---
In reply to: [276475053](https://github.com/dotnet/corefx/pull/36951#discussion_r276475053) [](ancestors = 276475053)

File: src/Microsoft.VisualBasic.Core/src/Microsoft/VisualBasic/CompilerServices/CharArrayType.vb [hunk 12]
```diff
@@ -0,0 +1,65 @@
...
+
+    <System.ComponentModel.EditorBrowsableAttribute(System.ComponentModel.EditorBrowsableState.Never)>
+    Public NotInheritable Class CharArrayType
+        ' Prevent creation.
+        Private Sub New()
+        End Sub
+
+        Public Shared Function FromString(ByVal Value As String) As Char()
+
+            If Value Is Nothing Then
```
[the remark this answers is not in the document]
Comment (cston) [reply src/Microsoft.VisualBasic.Core/src/Microsoft/VisualBasic/CompilerServices/CharArrayType.vb]: We should avoid changing the implementation unnecessarily.

---
In reply to: [276471070](https://github.com/dotnet/corefx/pull/36951#discussion_r276471070) [](ancestors = 276471070)

**Comment by cston** [comment]
@Anipik, having a copy of the same test data is useful for a couple of reasons: it means the test data can be located in the same class as the test; and it allows different tests to behave differently with the same inputs. (The second reason may not apply to the copy in this PR, but it does apply for the next PR.)

Perhaps I could consider merging test data later, if it still makes sense to do so.

**Pull request merged**